text/0000-non-exhaustive.md

```diff
@@ -0,0 +1,548 @@
+
+An attribute `#[non_exhaustive]` is added to the language, which will (for now)
+fail to compile if it's used on anything other than an enum, struct definition,
+or enum variant.
+
+## Enums
+
+Within the crate that defines the enum, this attribute is essentially ignored,
+so that the current crate can continue to exhaustively match the enum. The
+justification for this is that any changes to the enum will likely result in
+more changes to the rest of the crate. Consider this example:
+
+```rust
+use std::error::Error as StdError;
+
+#[non_exhaustive]
+pub enum Error {
+    Message(String),
+    Other,
+}
+impl StdError for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Message(ref s) => s,
+            Other => "other or unknown error",
+        }
+    }
+}
+```
+
+It seems undesirable for the crate author to use a wildcard arm here, to
+ensure that an appropriate description is given for every variant. In fact, if
+they use a wildcard arm in addition to the existing variants, it should be
+identified as dead code, because it will never be run.
+
+Outside the crate that defines the enum, users should be required to add a
+wildcard arm to ensure forward-compatibility, like so:
+
+```rust
+use mycrate::Error;
+
+match error {
+    Message(ref s) => ...,
+    Other => ...,
+    _ => ...,
+}
+```
+
+And it should *not* be marked as dead code, even if the compiler does mark it as
+dead and remove it.
+
+Note that this can *potentially* cause breaking changes if a user adds
+`#[deny(dead_code)]` to a match statement *and* the upstream crate removes the
+`#[non_exhaustive]` lint. That said, modifying warn-only lints is generally
+assumed to not be a breaking change, even though users can make it a breaking
+change by manually denying lints.
+
+## Structs
+
+Like with enums, the attribute is essentially ignored in the crate that defines
+the struct, so that users can continue to construct values for the struct.
+However, this will prevent downstream users from constructing or exhaustively
+matching the struct, because fields may be added to the struct in the future.
+
+Additionally, adding `#[non_exhaustive]` to an enum variant will operate exactly
+the same as if the variant were a struct.
+
+Using our `Config` again:
+
+```rust
+#[non_exhaustive]
+pub struct Config {
+    pub window_width: u16,
+    pub window_height: u16,
+}
+```
+
+We can still construct our config within the defining crate like so:
+
+```rust
+let config = Config { window_width: 640, window_height: 480 };
+```
+
+And we can even exhaustively match on it, like so:
+
+```rust
+if let Ok(Config { window_width, window_height }) = load_config() {
+    // ...
+}
+```
+
+But users outside the crate won't be able to construct their own values, because
+otherwise, adding extra fields would be a breaking change.
+
+Users can still match on `Config`s non-exhaustively, as usual:
+
+```rust
+let &Config { window_width, window_height, .. } = config;
```

Comment:
I didn't read the comments before reviewing, only the normative text.

Now I see the justification of prohibiting FRU:
>that's basically a promise not to add private fields in the future

I.e. a struct with all public fields won't be able to add a private field.
The silly FRU desugaring strikes again, this would be totally okay with the alternative desugaring.
Okay, strike FRU for now. (Even better, the RFC should describe why FRU is prohibited despite being compatible with non-exhaustiveness in theory.)